File: sdk/clawblox-ts/examples/basic_agent.ts
// Basic PPO-style agent loop example
import { ClawBloxAgent, ClawBloxClient, ObserveStateResponse } from '../src';

// Simple policy function - replace with your own
function policy(state: ObserveStateResponse): string {
  // Example: move forward
  return `game:GetService("Players").LocalPlayer.Character:MoveTo(Vector3.new(0, 0, -5))`;
}

async function main() {
  const agent = new ClawBloxAgent();
  
  // Get initial state
  let state = await agent.reset();
  
  console.log('Agent started with seed:', agent['seed']);
  
  for (let step = 0; step < 1000; step++) {
    // Get action from policy
    const action = policy(state);
    
    // Execute step
    const { state: nextState, done } = await agent.step(action);
    
    state = nextState;
    
    if (done) {
      console.log(`Episode done at step ${step}, resetting...`);
      state = await agent.reset();
    }
    
    if (step % 100 === 0) {
      console.log(`Step ${step}: tick = ${state.metadata?.tick}`);
    }
  }
  
  await agent.destroy();
  console.log('Agent destroyed');
}

main().catch(console.error);

File: sdk/clawblox-py/examples/basic_agent.py
#!/usr/bin/env python3
"""Basic agent example - Simple PPO-style loop with random actions."""

import random

from clawblox import ClawBloxAgent


def main():
    """Run a simple agent loop."""
    # Initialize agent with deterministic mode
    agent = ClawBloxAgent(deterministic=True, seed=42)

    # Reset to get initial state
    state = agent.reset()
    print(f"Initial state: {len(state.instances)} instances")

    # Define a set of possible actions
    actions = [
        'game.Workspace.Part.Position = Vector3.new(0, 5, 0)',
        'game.Workspace.Part.Size = Vector3.new(4, 4, 4)',
        'game.Workspace.Part.Color = Color3.new(1, 0, 0)',
        'game.Workspace.Part.Transparency = 0.5',
        'game.Workspace.Part.Anchored = true',
    ]

    # Run the agent loop
    total_reward = 0.0
    for step in range(1000):
        # Choose random action
        action = random.choice(actions)

        # Execute step
        result = agent.step(action)

        # Track reward
        total_reward += result.reward
        state = result.state

        # Print progress every 100 steps
        if (step + 1) % 100 == 0:
            print(f"Step {step + 1}: reward={result.reward:.4f}, total={total_reward:.4f}, instances={len(state.instances)}")

        # Reset if done
        if result.done:
            print(f"Resetting at step {step + 1}")
            state = agent.reset()

    # Clean up
    agent.destroy()
    print(f"Done! Total reward: {total_reward:.4f}")


if __name__ == "__main__":
    main()
